text/0000-newtype.md

```diff
@@ -0,0 +1,360 @@
+```c
+int Gen(long input);
+```
+
+In Rust the signature looks like this:
+
+```rust
+extern {
+    fn Gen(input: c_long) -> c_int;
+}
+```
+
+We create a rustic wrapper:
+
+```rust
+fn gen(input: i64) -> i32 {
+    unsafe { Gen(input) }
+}
+```
+
+This will work fine on our Linux development box but on windows the compilation
+will fail because `c_long` is `i32`.
+
+Correct, platform independent code will have to add explicit casts:
+
+```rust
+fn gen(input: i64) -> i32 {
+    unsafe { Gen(input as c_long) as i32 }
+}
+```
+
+There are currently some proposals regarding `as` that would worsen this
+situation:
+
+1. Make useless casts an error or warning: `input as c_long` would cause an
+   error or warning because `input` and `c_long` are both `i64`.
+2. Add implicit widening: This would add more cases where `as` causes a warning
+   or error.
+
+As an alternative, one might think about using newtype structs for `c_long`:
+
+```rust
+#[cfg(unix)]
+struct c_long(i64);
+#[cfg(windows)]
+struct c_long(i32);
+
+impl FromPrimitive for c_long { ... }
+```
+
+This solution will enforce explicit conversions, however, they can't be used
+with `as` and create a lot of friction in the code:
+
+```rust
+fn gen(input: i64) -> i32 {
+    unsafe { Gen(FromPrimitive::from_i64(input).unwrap()).0 as i32 }
+}
+```
+
+(Note that the error checking performed by `FromPrimitive` above is not always
+desired.)
+
+Below we propose an alternative that sits between `type` and newtype structs.
+
+# Detailed design
+
+Rename `type` to `alias` to better document its function.
+
+Add a `newtype` keyword.
+
+## Syntax
+
+```rust
+newtype T = U;
+```
+
+where `T` and `U` are as in the definition of `alias`.
+
+(You might want to first skip the details and continue with the "Rationale"
+section below.)
+
+## Semantics
+
+`T` behaves as if it had been declared by `struct T(());` except as described
+below.
+
+### Representation
+
+`T` has the same representation in memory as `U` if any.
+
+### Casting
+
+Let `R1` be the binary, transitive, symmetric, irreflexive relation generated by
+all `newtype` definitions, all `alias` definitions, and the numeric casts built
+into the language. 
+
+If `(T, U) \in R1` and `x: T`, then `x as U` is a valid cast.
+
+Let `R2` be the binary, transitive, symmetric, irreflexive relation generated by
+all `newtype` definitions and all `alias` definitions.
+
+Let `S` be the set of reference operators, `S = {&, &mut}`, and let `W` be a
+word generated by `S`, e.g., `W = e` or `W = &&&mut &&mut &`.
+
+If `(T, U) \in R2` and `x: WT`, then `x as WU` is a valid cast.
+
+### Inference
+
+Let `R2` be as above.
+
+If `U` is an integer type and `(T, U) \in R2`, then untyped integer literals will
+be infered as `T` in `T` position.
+
+If `U` is a floating point type and `(T, U) \in R2`, then untyped floating point
+literals will be infered as `T` in `T` position.
+
+### Traits
+
+No built-in traits are or can be implemented for `T` except as described below.
```

Comment:
That is a nebulous/imprecise and somewhat useless definition, e.g. `Send` and `Sync` will hopefully lose their `#[lang]` attributes, and `Clone` doesn't have an attribute but is a more useful/"fundamental" trait than many of the lang items.